### ISSUE TITLE
sft微调时报错

### ISSUE THREAD

**Issue by ama0zarashi** [charent/ChatLM-mini-Chinese#37]
![截屏2024-03-20 17 59 05](https://github.com/charent/ChatLM-mini-Chinese/assets/40602848/4efb6cb1-9e49-4c23-960d-147e708e29e4)
![截屏2024-03-20 18 02 54](https://github.com/charent/ChatLM-mini-Chinese/assets/40602848/45cc53fb-100b-46d8-bf9a-4685f7525837)
在sagemaker平台上进行sft微调，训练时出现爆显存的问题，以下是我的训练设置，就算换成24g显存的机器依然不够
![截屏2024-03-20 18 04 04](https://github.com/charent/ChatLM-mini-Chinese/assets/40602848/6e3c054d-0bdc-4b8e-99ea-b36584523326)
训练数据有135045条，每条prompt和response的token一共1500左右

**Comment by charent** [comment]
不应该啊，我这边16GB显存都可以微调。你使用的是哪个trainer？你现在的配置是`total batch_size = 32`，你设置`batch_size = 4`，`gradient_accumulation_steps = 8`试试。 



如果你使用huggingface trainer的话，参考这里：[sft.py#L159](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/train_model/sft.py#L159)和[utils.py#L9](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/utils/utils.py#L9)，加上清除cuda缓存的回调。如果使用本项目的trainer，在每个step结束后加上`torch.cuda.empty_cache()`试试。

**Comment by ama0zarashi** [comment]
> 不应该啊，我这边16GB显存都可以微调。你使用的是哪个trainer？你现在的配置是`total batch_size = 32`，你设置`batch_size = 4`，`gradient_accumulation_steps = 8`试试。
> 
> 如果你使用huggingface trainer的话，参考这里：[sft.py#L159](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/train_model/sft.py#L159)和[utils.py#L9](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/utils/utils.py#L9)，加上清除cuda缓存的回调。如果使用本项目的trainer，在每个step结束后加上`torch.cuda.empty_cache()`试试。

我试着使用本项目的sft_train.py进行训练，在每个step后加上`torch.cuda.empty_cache()`，还是老问题。我的配置如下

    max_seq_len: int = 384 + 8                # 8 for eos token 

    finetune_from_ckp_file = PROJECT_ROOT + '/model_save/'

    tokenizer_dir: str = PROJECT_ROOT + '/model_save/'  # tokenizer一般和model权重放在同一个文件夹
    sft_train_file: str = PROJECT_ROOT + '/data/sft_train.json'

    batch_size: int = 4
    num_train_epochs: int = 2
    save_steps: int = 5000
    gradient_accumulation_steps: int = 2
    learning_rate: float = 1e-5
    logging_first_step: bool = True
    logging_steps: int = 100
    output_dir: str = PROJECT_ROOT + '/model_save/sft'
    warmup_steps: int = 100
    fp16: bool = True
    seed: int = 23333

**Comment by ama0zarashi** [comment]
> 不应该啊，我这边16GB显存都可以微调。你使用的是哪个trainer？你现在的配置是`total batch_size = 32`，你设置`batch_size = 4`，`gradient_accumulation_steps = 8`试试。
> 
> 如果你使用huggingface trainer的话，参考这里：[sft.py#L159](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/train_model/sft.py#L159)和[utils.py#L9](https://github.com/charent/Phi2-mini-Chinese/blob/75210853cab1d4a4dfd7914630a331a12c39ea7f/utils/utils.py#L9)，加上清除cuda缓存的回调。如果使用本项目的trainer，在每个step结束后加上`torch.cuda.empty_cache()`试试。

我是直接用本项目的sft_train.py进行训练，在回调函数里设置on_step_end，在每个step后用`torch.cuda.empty_cache()`清除缓存，设置`batch_size = 4`，`gradient_accumulation_steps = 8`的时候在训练到第33步时报错，我设置`batch_size = 4`，`gradient_accumulation_steps = 2`时，在训练到第132步时报错。都是batchsize降为0了，我将数据集的数量变为400后就没有这个问题了

**Comment by charent** [comment]
那可能是你的数据集有问题，检查一下是不是有空数据。

**Comment by ama0zarashi** [comment]
> 那可能是你的数据集有问题，检查一下是不是有空数据。

嗯，找到原因了，好像是单条数据的token太长了，我做了个截断后就不会爆显存了，谢谢大佬。
